src/lib/plugin.js

```diff
@@ -29,73 +29,107 @@ module.exports = class PluginVirtual extends EventEmitter2 {
   constructor (opts) {
     super()
 
-    assertOptionType(opts, 'currencyCode', 'string')
-    assertOptionType(opts, 'currencyScale', 'number')
-    assertOptionType(opts, 'maxBalance', 'string')
-    assertOptionType(opts, 'secret', 'string')
-    assertOptionType(opts, 'peerPublicKey', 'string')
-    assertOptionType(opts, '_store', 'object')
-    assertOptionType(opts, 'rpcUri', 'string')
-
-    this._secret = opts.secret
-    this._peerPublicKey = opts.peerPublicKey
-    this._publicKey = Token.publicKey(this._secret)
-    this._currencyCode = opts.currencyCode.toUpperCase()
-    this._currencyScale = opts.currencyScale
+    this._stateful = !!opts._store
+    if (this._stateful) {
+      assertOptionType(opts, 'currencyCode', 'string')
+      assertOptionType(opts, 'currencyScale', 'number')
+      assertOptionType(opts, 'maxBalance', 'string')
+
+      this._store = opts._store
+      this._maxBalance = opts.maxBalance
+      this._minBalance = opts.minBalance
+      this._highestBalance = new Balance({
+        key: 'balance__',
+        store: this._store,
+        maximum: this._maxBalance
+      })
+
+      this._lowestBalance = new Balance({
+        key: 'balance_l',
+        minimum: this._minBalance,
+        store: this._store
+      })
+
+      // give a 'balance' event on balance change
+      this._highestBalance.on('balance', (balance) => {
+        this.emit('balance', balance)
+      })
+    }
 
-    this._store = opts._store
-    this._maxBalance = opts.maxBalance
-    this._balance = new Balance({
-      maximum: this._maxBalance,
-      store: this._store
-    })
+    if (opts.rpcUris) {
+      assertOptionType(opts, 'rpcUris', 'object')
+      this._rpcUris = opts.rpcUris
+    } else {
+      assertOptionType(opts, 'rpcUri', 'string')
+      this._rpcUris = [ opts.rpcUri ]
+    }
 
-    // give a 'balance' event on balance change
-    this._balance.on('balance', (balance) => {
-      this.emit('balance', balance)
-    })
+    this._currencyCode = opts.currencyCode.toUpperCase()
+    this._currencyScale = opts.currencyScale
 
-    // Token uses ECDH to get the ledger prefix from secret and public key
-    this._authToken = Token.authToken({
-      secretKey: this._secret,
-      peerPublicKey: this._peerPublicKey
-    })
+    if (opts.token && opts.prefix) {
+      assertOptionType(opts, 'token', 'string')
+      assertOptionType(opts, 'prefix', 'string')
 
-    this._prefix = Token.prefix({
-      secretKey: this._secret,
-      peerPublicKey: this._peerPublicKey,
-      currencyCode: this._currencyCode,
-      currencyScale: this._currencyScale
-    })
+      this._peerAccountName = this._stateful ? 'client' : 'server'
+      this._accountName = this._stateful ? 'server' : 'client'
+      this._authToken = opts.token
+      this._prefix = opts.prefix
+    } else {
+      // deprecate this version?
```

Comment:
@michielbdejong had a good argument against deprecating this when you actually have two equal peers. If you don't use Diffie-Hellman you need to decide who's going to generate the token.

That said, it definitely wouldn't be terrible to get rid of the dependency on ed25519...